### ISSUE TITLE
javacpp presets for tensorflow distributed_runtime

### ISSUE THREAD

**Issue by debasish83** [bytedeco/javacpp-presets#398]
With SparkNet + TensorFrame using javacpp presets focusing on data parallel model training but to run model parallel training using spark/scala, we need tensorflow distributed_runtime to be exposed as part of javacpp presents / JNI...Is someone working on it ? I could not find an issue

**Comment by saudet** [comment]
Not that I know. If you would like to give it a try, it would be very appreciated! I'll help you out with issues you encounter.

**Comment by debasish83** [comment]
Yes I would like to try it out...do you know if protobuf etc can cause any issue...I doubt since protobuf are available for both java and c++....should I start with tensorflow presets ?

**Comment by saudet** [comment]
Yes, please do. Like I said, if you encounter any problems, let me know, and I will help.

**Comment by debasish83** [comment]
Do you know why tensorflow went to jni route for java interface and did not prefer to take swig / javacpp route ? Is there a specific benefit you see over using JNI compared to Javacpp ?

**Comment by debasish83** [comment]
swig bindings are there for most of the code as well that python uses and it could have been modified to export jni...

**Comment by debasish83** [comment]
I looked into tensorflow/python/training/server_lib.i which is the swig file for distributed_runtime python exposure and based on that added these headers:
"tensorflow/c/tf_status_helper.h",
"tensorflow/core/distributed_runtime/server_lib.h",
"tensorflow/core/lib/core/status.h",
"tensorflow/core/protobuf/tensorflow_server.pb.h"
I believe these are sufficient to do distributed training. I will look further into tensorflow.i swig file and try to expose them through javacpp-presets...
I will go through documentation but do I need add something more into InfoMapper as well ?

**Comment by saudet** [comment]
You might not need to add anything more, but it's not usually the case. Start by adding the header files, one by one, and try to fix the errors as they come up!

As to why they gave up using SWIG for Java, it's probably for the same reasons that I came up with JavaCPP, but I suppose they don't feel that JavaCPP is mature enough yet. It's too bad really that very few people are interested in solving this problem. Everyone agrees that it's an important problem, but it's not an interesting one to fix, apparently. :(

**Comment by debasish83** [comment]
Based on tensorflow.i these are headers exposed in python api. I will open up a PR after further tests but looks like javacpp presets for tensorflow has more functions compared to what python api exposes:
"tensorflow/core/distributed_runtime/server_lib.h",
"tensorflow/core/lib/core/status.h",
"tensorflow/core/protobuf/tensorflow_server.pb.h",
"tensorflow/c/checkpoint_reader.h",
"tensorflow/core/util/stat_summarizer.h",
"tensorflow/core/util/events_writer.h",
"tensorflow/core/util/event.pb.h",
"tensorflow/core/lib/core/errors.h",
"tensorflow/core/public/version.h",
"tensorflow/core/common_runtime/device.h",
"tensorflow/core/common_runtime/device_factory.h",
"tensorflow/core/framework/device_attributes.pb.h",
"tensorflow/core/graph/quantize_training.h",
"tensorflow/core/lib/core/stringpiece.h",
"tensorflow/core/lib/io/buffered_inputstream.h",
"tensorflow/core/lib/io/inputstream_interface.h",
"tensorflow/core/lib/io/random_inputstream.h",
"tensorflow/core/lib/io/path.h",
"tensorflow/core/protobuf/meta_graph.pb.h"

**Comment by debasish83** [comment]
Another question....When I worked on conic optimization to jni, I found that unless the critical section is used, there is significant runtime in copying arrays out from java to c++ before compute can happen in c++..do you use critical sections in javacpp ?

**Comment by saudet** [comment]
Not for normal function calls, it's not a good default, and direct NIO
buffers are even more efficient anyway. But it could be added as an option.
There's a RFE open if you'd like to make a contribution.

**Comment by debasish83** [comment]
Not for function calls but how about some blas / tensor style calls which sends a f(a: Array[Double], b: Array[Double], c: Double) and then operation is a += c*b...there must be similar variants over tensor

**Comment by saudet** [comment]
Yeah, that's what the Pointer class is for.

**Comment by saudet** [comment]
BTW, the reason they can use netlib-java in Spark is because there is a pure Java backend available via F2J. However, F2J does not support direct NIO buffers. If they could easily switch to direct NIO buffers they would have already, but it would require a lot of effort: https://github.com/fommil/netlib-java/issues/58

There are otherwise no reasons to rely on Java arrays. Just use direct NIO buffers, which are both faster and safer than critical sections on Java arrays.
